proxy/src/pglb/inprocess.rs

```diff
@@ -0,0 +1,465 @@
+pub fn create_server_endpoint(
+    server_socket: InProcessUdpSocket,
+    cert: pki_types::CertificateDer<'static>,
+    key: pki_types::PrivateKeyDer<'static>,
+) -> anyhow::Result<quinn::Endpoint> {
+    let mut server_crypto = rustls::ServerConfig::builder()
+        .with_no_client_auth()
+        .with_single_cert(vec![cert], key)?;
+    server_crypto.alpn_protocols = vec![b"pglb".into()];
+    server_crypto.key_log = Arc::new(rustls::KeyLogFile::new());
+
+    let mut server_config =
+        quinn::ServerConfig::with_crypto(Arc::new(QuicServerConfig::try_from(server_crypto)?));
+    let transport_config = Arc::get_mut(&mut server_config.transport).expect("no other clone");
+    transport_config.max_concurrent_uni_streams(0_u8.into());
+
+    let server = quinn::Endpoint::new_with_abstract_socket(
+        quinn::EndpointConfig::default(),
+        Some(server_config),
+        Arc::new(server_socket),
+        Arc::new(quinn::TokioRuntime),
+    )?;
+
+    Ok(server)
+}
+
+pub fn create_client_endpoint(
+    client_socket: InProcessUdpSocket,
+    ca: pki_types::CertificateDer<'static>,
+) -> anyhow::Result<quinn::Endpoint> {
+    let mut client = quinn::Endpoint::new_with_abstract_socket(
+        quinn::EndpointConfig::default(),
+        None,
+        Arc::new(client_socket),
+        Arc::new(quinn::TokioRuntime),
+    )?;
+
+    let mut root_certs = rustls::RootCertStore::empty();
+    root_certs.add(ca)?;
+
+    let mut client_crypto = rustls::ClientConfig::builder()
+        .with_root_certificates(root_certs)
+        .with_no_client_auth();
+    client_crypto.alpn_protocols = vec![b"pglb".into()];
+    client_crypto.key_log = Arc::new(rustls::KeyLogFile::new());
+
+    let client_config =
+        quinn::ClientConfig::new(Arc::new(QuicClientConfig::try_from(client_crypto)?));
+    client.set_default_client_config(client_config);
+
+    Ok(client)
+}
+
+struct Datagram<const MSS: usize> {
+    len: usize,
+    buf: [u8; MSS],
+}
+
+struct DatagramQueue<const MSS: usize> {
+    sender: TryLock<CachingProd<Arc<HeapRb<Datagram<MSS>>>>>,
+    receiver: TryLock<CachingCons<Arc<HeapRb<Datagram<MSS>>>>>,
+    receiver_waker: Arc<ReceiverWaker>,
+}
+
+impl<const MSS: usize> DatagramQueue<MSS> {
+    fn new(
+        sender: CachingProd<Arc<HeapRb<Datagram<MSS>>>>,
+        receiver: CachingCons<Arc<HeapRb<Datagram<MSS>>>>,
+    ) -> Self {
+        Self {
+            sender: TryLock::new(sender),
+            receiver: TryLock::new(receiver),
+            receiver_waker: Arc::new(ReceiverWaker::new()),
+        }
+    }
+
+    #[tracing::instrument(level = "debug", skip_all, err, fields(packet = packet.len()))]
+    fn send(&self, packet: &[u8]) -> io::Result<()> {
+        if packet.len() > MSS {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "send: packet too large",
+            ));
+        }
+
+        let mut sender = self.sender.try_lock().expect("no other thread");
```

Comment:
Are we sure that try_lock is sufficient here? Multiple quinn Streams might be writing to the same InProcessUdpSocket/DatagramQueue at a time.

Reply:
Good point. No, it's actually not here. The individual SendStreams will directly use the socket to send data. I changed this to a std Mutex now.